core/Context.cc

```diff
@@ -117,26 +117,44 @@ GlobalSubstitution::GlobalSubstitution(const GlobalState &from, GlobalState &to,
     if (!fastPath || debug_mode) {
         {
             UnfreezeNameTable unfreezeNames(to);
-            nameSubstitution.reserve(from.names.size());
+            utf8NameSubstitution.reserve(from.utf8Names.size());
+            constantNameSubstitution.reserve(from.constantNames.size());
+            uniqueNameSubstitution.reserve(from.uniqueNames.size());
             int i = -1;
-            for (const Name &nm : from.names) {
+            for (const UTF8Name &nm : from.utf8Names) {
                 i++;
-                ENFORCE(nameSubstitution.size() == i, "Name substitution has wrong size");
-                switch (nm.kind) {
-                    case NameKind::UNIQUE:
-                        nameSubstitution.emplace_back(to.freshNameUnique(
-                            nm.unique.uniqueNameKind, substitute(nm.unique.original), nm.unique.num));
-                        break;
-                    case NameKind::UTF8:
-                        nameSubstitution.emplace_back(to.enterNameUTF8(nm.raw.utf8));
-                        break;
-                    case NameKind::CONSTANT:
-                        nameSubstitution.emplace_back(to.enterNameConstant(substitute(nm.cnst.original)));
-                        break;
-                    default:
-                        ENFORCE(false, "NameKind missing");
+                ENFORCE_NO_TIMER(utf8NameSubstitution.size() == i, "UTF8 name substitution has wrong size");
+                utf8NameSubstitution.emplace_back(to.enterNameUTF8(nm.utf8));
+                ENFORCE(!fastPath || utf8NameSubstitution.back().utf8Index() == i);
+            }
+            // UniqueNames and ConstantNames may reference each other, necessitating some special logic here to avoid
+            // crashing. We process UniqueNames first because there are fewer of them, so fewer loop iterations require
+            // this special check. Tested in `core_test.cc`.
+            i = -1;
+            for (const UniqueName &nm : from.uniqueNames) {
+                i++;
+                ENFORCE(uniqueNameSubstitution.size() == i, "Unique name substitution has wrong size");
+                if (nm.original.kind() == NameKind::CONSTANT &&
+                    nm.original.constantIndex() >= constantNameSubstitution.size()) {
+                    for (u4 i = constantNameSubstitution.size(); i <= nm.original.constantIndex(); i++) {
+                        auto &cnst = from.constantNames[i];
+                        ENFORCE_NO_TIMER(constantNameSubstitution.size() == i,
+                                         "Constant name substitution has wrong size");
+                        // N.B.: cnst may reference a UniqueName, but since names are linearizeable we should have
+                        // already substituted it by now.
+                        constantNameSubstitution.emplace_back(to.enterNameConstant(substitute(cnst.original)));
```

Comment:
It's a little unfortunate that this loop body gets duplicated below. It's probably worth adding a warning that changing one will require changing the other.

Reply:
done